File: BOJ_prob/DFS_and_BFS/python/BOJ_1991.py
# A: 65 Z: 90 . : 46
import sys

def preorder(c_node):
    print(chr(c_node + 65), end = '')
    for i in range(2):
        if adj_list[c_node][i] != 46:
            preorder(adj_list[c_node][i] - 65)

def inorder(c_node):
    for i in range(2):
        if adj_list[c_node][i] != 46:
            inorder(adj_list[c_node][i] - 65)
        if i == 0:
            print(chr(c_node + 65), end = '')

def postorder(c_node):
    for i in range(2):
        if adj_list[c_node][i] != 46:
            postorder(adj_list[c_node][i] - 65)
    print(chr(c_node + 65), end = '')


nodenum = int(sys.stdin.readline())
adj_list = [[] for i in range(nodenum)]
for i in range(nodenum):
    treeinfo = list(map(ord, sys.stdin.readline().split()))
    for i in range(1, 3):
        adj_list[int(treeinfo[0]) - 65].append(treeinfo[i])

preorder(0)
print()
inorder(0)
print()
postorder(0)
    

File: BOJ_prob/DFS_and_BFS/C++/BOJ_1991.cpp
#include <iostream>
#include <vector>


using namespace std;

vector<char> tree[26];

void preOrder(char cur_node);
void inOrder(char cur_node);
void postOrder(char cur_node);

int main()
{
    int node_num;
    cin >> node_num;
    for(int i = 0; i < node_num; i++){
        char parent, lc, rc;
        cin >> parent >> lc >> rc;
        tree[parent - 'A'].push_back(lc);
        tree[parent - 'A'].push_back(rc);
    }
    preOrder('A');
    printf("\n");
    inOrder('A');
    printf("\n");
    postOrder('A');
}

void preOrder(char cur_node)
{
    printf("%c", cur_node);
    if(tree[cur_node - 'A'][0] != '.')
        preOrder(tree[cur_node - 'A'][0]);
    if(tree[cur_node - 'A'][1] != '.')
        preOrder(tree[cur_node - 'A'][1]);
}

void inOrder(char cur_node)
{
    if(tree[cur_node - 'A'][0] != '.')
        inOrder(tree[cur_node - 'A'][0]);
    printf("%c", cur_node);
    if(tree[cur_node - 'A'][1] != '.')
        inOrder(tree[cur_node - 'A'][1]);
}

void postOrder(char cur_node)
{
    if(tree[cur_node - 'A'][0] != '.')
        postOrder(tree[cur_node - 'A'][0]);
    if(tree[cur_node - 'A'][1] != '.')
        postOrder(tree[cur_node - 'A'][1]);
    printf("%c", cur_node);
}